### PR TITLE
avoid unapply due to issue reported by Play team (scala 3 only)

### PR DESCRIPTION
**Pull request by pjfanning**
* relates to https://github.com/apache/incubator-pekko-http/issues/183
* simplifies the code, we don't really need the unapply

### CODE REVIEW

**Comment by mkurz** [comment]
Nice that was fast :+1:

**Review by mkurz — APPROVED** [review]
LGTM

**Comment by mdedetrich** [comment]
Nice work!

Is it possible to write a small unit test that fails before this change and passes with this change?

**Comment by pjfanning** [comment]
> Nice work!
> 
> Is it possible to write a small unit test that fails before this change and passes with this change?

I added https://github.com/apache/incubator-pekko-http/pull/185 to see if any tests went through this code. Plenty do.
I suspect that something more is going on.

@mkurz mentioned that where he saw the issue was in sbt scripted tests. Maybe sbt could be doing something like instrumenting the classes under test.

**Comment by mdedetrich** [comment]
Well since this is a compiler change I am not that fussed about a unit test for this specific part of the code, I just suggested it because its easier.

What would be more ideal is a normal/integration test that happens to cover this, i.e. we should be testing this part of the code one way or another.

**Review by mdedetrich — APPROVED** [review]
lgtm

**Comment by pjfanning** [comment]
The code is actually well tested. So other than adding the sbt scripted test from the Play Framework, I am not really sure how to extend the tests here.

This is fairly minor refactor, since the unapply is not really necessary. We already have the match case matching to a variable called `first`.

**Comment by mdedetrich** [comment]
> The code is actually well tested. So other than adding the sbt scripted test from the Play Framework, I am not really sure how to extend the tests here.
> 
> This is fairly minor refactor, since the unapply is not really necessary. We already have the match case matching to a variable called `first`.

Well unless I am missing something, evidently we dont have a test that goes through this code path because otherwise we would have hit it. Or is the case that we aren't running tests with Scala 3 specifically (in which case we should?)

**Comment by pjfanning** [comment]
We have lots of tests going through the code and we do test with Scala 3.

With sbt scripted tests, there could be asm manipulation of the classes or something like that. I had to investigate an issue in Jackson recently where a user found a problem when using Gradle with Jackson and it turned out that Gradle was using asm and the issue was stemming from there.

**Comment by mdedetrich** [comment]
> With sbt scripted tests, there could be asm manipulation of the classes or something like that.

I don't think this is the case, sbt-scripted just compiles the code and generates the jar just like a normal user would and just runs the code, in essence its closer to a smoke test. As a hypothetical example if we had inlining in pekko-http that could be an example, i.e. inlining only enabled on CI/build but not when running normally.

My hunch is honestly that we are not going through this code path, but looks like you are already on it here https://github.com/apache/incubator-pekko-http/pull/185

**Comment by He-Pin** [comment]
A Scala 3 bug?

File: http-core/src/main/scala/org/apache/pekko/http/impl/engine/http2/ProtocolSwitch.scala [hunk 1]
```diff
@@ -62,11 +62,12 @@ private[http] object ProtocolSwitch {
               new InHandler {
                 def onPush(): Unit =
                   grab(netIn) match {
-                    case first @ SessionBytes(session, bytes) =>
+                    case first: SessionBytes =>
```
Comment (He-Pin) [inline http-core/src/main/scala/org/apache/pekko/http/impl/engine/http2/ProtocolSwitch.scala:65]: lgtm, fire an issue in dotty too?

**Review by He-Pin — APPROVED** [review]
lgtm

**Comment by pjfanning** [comment]
> A Scala 3 bug?

Probably - but I haven't been able to come up with a minimised test case that could be reported to Dotty team.

**Pull request merged**